### PR TITLE
WIP: Implement intention and inspection for let mut

### PR DESCRIPTION
**Pull request by farodin91**
Fixes partly #1004 and #1005.
Done
- Add error annotator for immutable variable 
  - callArg
  - methodCall
  - self
- Improved error message for example ```"Cannot borrow immutable local variable `${expr.text}` as mutable"```
- Add mutable as quick fix (annotator)
- Add Intention for `Add mutable` and `Remove mutable` **(WIP)** only for let and should use the same update code as quick fix for annotator
- Add inspection for No required mutable variable **(WIP)**

ToDo
- tests

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    return visitor.foundOccurrences
+}
+
+class RsVariableMutableInspection : RsLocalInspectionTool() {
+    override fun getDisplayName() = "No mutable required"
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) =
+        object : RsVisitor() {
+            override fun visitPatBinding(o: RsPatBinding) {
+                if (!o.isMut) return
+                val occurrences = findOccurrences(o.identifier)
```
Comment (farodin91) [inline src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt]: Is their a better way to get all occurrences, like in rename.

File: src/main/kotlin/org/rust/ide/intentions/RemoveMutableIntention.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    open val mutable = true
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val newLetExpr = RsPsiFactory(project).createLetDeclaration(ctx.letExpr.pat!!.firstChild.lastChild.text, ctx.letExpr.expr!!, !mutable, ctx.letExpr.typeReference)
+        ctx.letExpr.replace(newLetExpr)
+    }
+
+    protected fun isMutable(letExpr: RsLetDecl): Boolean {
+        val pat = letExpr.pat?.firstChild?.firstChild ?: return false
```
Comment (farodin91) [inline src/main/kotlin/org/rust/ide/intentions/RemoveMutableIntention.kt]: Any tip clean up this?

File: src/main/kotlin/org/rust/ide/annotator/fixes/AddMutableFix.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    override fun getFamilyName(): String = text
+    override fun getText(): String  = "Add mutable"
+    open val mutable = true
+
+    override fun invoke(project: Project, file: PsiFile, editor: Editor?, startElement: PsiElement, endElement: PsiElement) {
+        val left = binaryExpr.left
+        val path = (left as? RsPathExpr)?.path ?: return
+        val declaration = path.reference.resolve()
+        val letExpr = declaration!!.parentOfType<RsLetDecl>()!!
+        val newLetExpr = RsPsiFactory(project)
```
Comment (farodin91) [inline src/main/kotlin/org/rust/ide/annotator/fixes/AddMutableFix.kt]: Any tip to cleanup this? This my first contact with Kotlin.

**Comment by farodin91** [comment]
@matklad Do you like to look into my current state?

File: src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt (hunk 1 above)
Comment (matklad) [inline src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt]: You need something different than occurrences here, you need `ReferencesSearch`. See how it is used in `org.rust.ide.annotator.RsTraitLineMarkerProvider` .

File: src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                    for (occurrence in occurrences) {
+                        val parent = occurrence.parentOfType<RsBinaryExpr>() ?: continue
+                        if (parent.left == occurrence) {
+                            isMutableNeeded = true
+                            break
+                        }
+                    }
+                }
+                if (!isMutableNeeded) {
+                    holder.registerProblem(o, "No mutable required for ${o.identifier.text}", RemoveMutableFix(o))
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt]: Hm, I think this'll give false positives for stuff like

```Rust
struct S; 
impl S {  fn needs_mut(&mut self) {} }

fn main() {
  let s = S;
  s.needs_mut()
}
```


I am not actually sure that we can implement this inspection without false positives reliably at the moment...

File: src/main/kotlin/org/rust/ide/intentions/RemoveMutableIntention.kt (hunk 2 above)
Comment (matklad) [inline src/main/kotlin/org/rust/ide/intentions/RemoveMutableIntention.kt]: I think you want to exactly switch on the type of possible patterns. Remember, you can have stuff like `let (foo, bar) = (1, 2)`.

**Comment by matklad** [comment]
Looks nice! But let's tread lightly and implement this piece by piece!

File: src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt (hunk 4 above)
Comment (farodin91) [inline src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt]: This is only started currently

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 5]
```diff
@@ -51,11 +51,31 @@ class RsErrorAnnotator : Annotator, HighlightRangeExtension {
...
+        args
+            .filter { it.expectedMutable() && !it.isMutable() }
+            .map { (it as? RsUnaryExpr)?.expr }
+            .filter { it != null }
+            .forEach {
+                createImmutableErrorAnnotation(holder, it!!)
+            }
+    }
+
+    private fun checkMethodCall(holder: AnnotationHolder, expr: RsMethodCallExpr) {
```
Comment (farodin91) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: I use this for self.

**Comment by farodin91** [comment]
My error annotator implementation do the most stuff. I could move intention and inspection into a new PR.

**Comment by matklad** [comment]
>I could move intention and inspection into a new PR.

Yes, that's exactly what I have in mind, thanks!

**Comment by farodin91** [comment]
I'll close this and open new PRs.

**Pull request closed without merging**